### PR TITLE
[Feature] ORCID login [#OSF-5162, 6881, 6885]

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
Allow users to login via ORCID's OAuth.

## Changes
* New `DictionaryField` on `User` to store external auth ID's
* Custom login flow
* Login template changes
* Tests
* Email templates for this auth process
* Update keys from CAS

Sample CAS Response:
```
<cas:serviceResponse xmlns:cas='http://www.yale.edu/tp/cas'>
    <cas:authenticationSuccess>
        <cas:user>OrcidProfile#8733-8493-5068-5235</cas:user>
        <cas:attributes>
            <cas:given-names>Carolynn</cas:given-names>
            <cas:isFromNewLogin>true</cas:isFromNewLogin>
            <cas:authenticationDate>Tue Aug 23 00:09:02 EDT 2016</cas:authenticationDate>
            <cas:longTermAuthenticationRequestTokenUsed>false</cas:longTermAuthenticationRequestTokenUsed>
            <cas:accessToken>AT-1-0838eb2b58f4b6b0ecfdf7bd63f521447e3ef141a987f64c27df7beb4ba11341</cas:accessToken>
            <cas:family-name>Emmerich</cas:family-name>
        </cas:attributes>
    </cas:authenticationSuccess>
</cas:serviceResponse>
```
## Side effects
None expected

## Ticket
[OSF-6885](https://openscience.atlassian.net/browse/OSF-6885)
[OSF-6881](https://openscience.atlassian.net/browse/OSF-6881)
[OSF-5162](https://openscience.atlassian.net/browse/OSF-5162)

### CODE REVIEW

File: framework/auth/__init__.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
     'get_user',
     'check_password',
     'authenticate',
+    'oauth_first_time_authenticate',
```
Comment (mfraezz) [inline framework/auth/__init__.py]: This probably needs to be updated.

File: framework/auth/views.py [hunk 2]
```diff
@@ -637,3 +714,114 @@ def resend_confirmation_post(auth):
...
+    form = ResendConfirmationForm(request.form)
+    session = get_session()
+    if not session.is_external_first_login:
+        raise HTTPError(http.UNAUTHORIZED)
+
+    external_id_provider = session.data['auth_user_external_id_provider']
+
+    return {
+        'form': form,
+        'external_id_provider': external_id_provider.upper()
```
Comment (mfraezz) [inline framework/auth/views.py]: This shouldn't be calling `.upper()`

Comment (cslzchen) [inline framework/auth/views.py]: Agreed. `external_id_provider` used to be all lower-case and this `.upper()` is only for **ORCID** . With the new PR, it becomes what they should be in default settings for different external identity providers. There are other places that I used `.upper()` and I will fix all.

**Comment by cslzchen** [comment]
Related CAS PR: https://github.com/CenterForOpenScience/cas-overlay/pull/22.

File: framework/auth/cas.py [hunk 3]
```diff
@@ -264,7 +264,7 @@ def make_response_from_ticket(ticket, service_url):
             user = {
                 'external_id_provider': external_credential['provider'],
                 'external_id': external_credential['id'],
-                'fullname': cas_resp.attributes['given-names'] + ' ' + cas_resp.attributes['family-name'],
+                'fullname': '{} {}'.format(cas_resp.attributes.get('given-names'), cas_resp.attributes.get('family-name')),
```
Comment (cslzchen) [inline framework/auth/cas.py]: Just FYI: If both names are empty, use the ORCiD id instead. (HotFixed)

Comment (mfraezz) [inline framework/auth/cas.py]: When testing on the ORCiD website, they seemed to disallow an empty `given-names`.

Comment (cslzchen) [inline framework/auth/cas.py]: User can make their profile private and only ORCiD is returned. This is a very rare case we happened to ran into. :)

**Pull request merged**